kakao/KakaoLoginServer/app.js

```diff
@@ -33,47 +36,71 @@ function requestMe(kakaoAccessToken) {
     headers: {'Authorization': 'Bearer ' + kakaoAccessToken},
     url: requestMeUrl,
   });
-};
+}
+
+/**
+ * validateToken - Returns access token info from Kakao API,
+ * which checks if this token is issued by this application.
+ *
+ * @param {String} kakaoAccessToken Access token retrieved by Kakao Login API
+ * @return {Promise<Response>}      Access token info response
+ */
+function validateToken(kakaoAccessToken) {
+  console.log('Validating access token from Kakao API server.');
+  return request({
+      method: 'GET',
+      headers: {'Authorization': 'Bearer ' + kakaoAccessToken},
+      url: accessTokenInfoUrl,
+  });
+}
 
 
 /**
- * updateOrCreateUser - Update Firebase user with the give email, create if
+ * updateOrCreateUser - Update Firebase user with the give email, or create if
  * none exists.
  *
  * @param  {String} userId        user id per app
  * @param  {String} email         user's email address
  * @param  {String} displayName   user
  * @param  {String} photoURL      profile photo url
- * @return {Prommise<UserRecord>} Firebase user record in a promise
+ * @return {Promise<UserRecord>} Firebase user record in a promise
  */
 function updateOrCreateUser(userId, email, displayName, photoURL) {
-  console.log('updating or creating a firebase user');
-  const updateParams = {
-    provider: 'KAKAO',
-    displayName: displayName,
-  };
-  if (displayName) {
-    updateParams['displayName'] = displayName;
-  } else {
-    updateParams['displayName'] = email;
-  }
-  if (photoURL) {
-    updateParams['photoURL'] = photoURL;
-  }
-  console.log(updateParams);
-  return firebaseAdmin.auth().updateUser(userId, updateParams)
-  .catch((error) => {
-    if (error.code === 'auth/user-not-found') {
-      updateParams['uid'] = userId;
-      if (email) {
-        updateParams['email'] = email;
+  console.log(`fetching a firebase user by email ${email}`);
+  return firebaseAdmin.auth().getUserByEmail(email)
+    .then((userRecord) => linkUserWithKakao(userId, userRecord))
```

Comment:
Here you should check that the user is a Kakao user by check ing that he has an attribute `lineUID` equals to the Kakao UID.

Then if the user is already a Kakao user, no need to link him again. Just return the user.

If the user is existing but he is NOT a Kakao user you need to "link" the user. Normally you should ask the user to sign-in using the other provider first (e.g. Google or Facebook) and only once he signed-in you can mark him as "linked".

Reply:
@nicolasgarnier 

How about utilizing email_verified field that most oauth providers provide? This can simplify login process since re-authenticating with facebook or google is not necessary.

Here is the modified logic.

- First check with uid kakao:${userId}
  - This should be done first. Users signed up with kakao should be able to login again whether he or she has email field or not.

- What if email is empty?
  - Email could be empty due to lack of user consent. There is no way to map this user with any existing user if email is not given. Custom provider UIDs cannot be queried from custom user claims. There are two choices.
    - Disallow creating a user
      - Users will have to re-authenticate with providers with valid email and link again by passing firebase ID token (which will be used to retrieve firebase uid on the server-side) and Kakao access token.
     - Just create a new user and deal with mapping later on user requests
  - In this sample, just go with the first choice, cuz it's just sample. :)

3. If  email is present and there is a user with the email, give custom token to user if email is verified, make users re-authenticate with other providers if not verified.